ruby/ql/lib/codeql/ruby/dataflow/internal/DataFlowDispatch.qll

```diff
@@ -1519,16 +1541,26 @@ predicate parameterMatch(ParameterPosition ppos, ArgumentPosition apos) {
   (ppos.isHashSplat() or ppos.isSynthHashSplat()) and
   (apos.isHashSplat() or apos.isSynthHashSplat())
   or
-  exists(int pos |
+  exists(int pos, boolean hasActualSplatParam, boolean hasActualSplatArg |
     (
-      ppos.isSplat(pos)
+      ppos.isSplat(pos) and
+      hasActualSplatParam = true // allow matching with synthetic splat argument
```

Comment:
This doesn't seem right. The restriction on `hasActualSplatParam` has no effect (other than satisfying binding checks), as the other conjunct (the one that binds `apos` further down) doesn't use `hasActualSplatParam` for anything.

To do what the comment is suggesting we'd have to add `hasActualSplatArg = true`, i.e. the caller must have an actual splat argument.

But overall I have to say this clause is incredibly hard to read. I think it could be greatly simplified by rewriting it in DNF and accept small amounts of duplication in favour of ease of understanding.

Reply:
Right, I'll restructure this.